docs/rfcs/027-crash-consistent-layer-map-through-index-part.md

```diff
@@ -0,0 +1,219 @@
+  * if we crash before index part PUT, local layer files will be deleted
+
+## Trade-Offs
+
+`disk_consistent_lsn` can go backwards across restarts if we crash before we've finished the index part PUT.
+Nobody should care about it, because the only thing that matters is `remote_consistent_lsn`.
+Compute certainly doesn't care about `disk_consistent_lsn`.
+
+If we crash before finishing the index part PUT, we have to re-do the work during restart:
+* wal ingest: losing not-yet-uploaded L0s; load on the **safekeepers** + work for pageserver
+* compaction: we lose the entire compaction iteration work; need to re-do it again
+* gc: no change to what we have today
+
+In comparison to the current reconciliation process, this means we will sometimes throw away needless work that was previously conserved.
+For example, if we crash after doing all the local changes but before _any_ remote S3 changes.
+Note however that the amount of work to be re-done after restart is capped to the lag of S3 changes to the local changes.
+That is something we can control and monitor for.
+
+Related, if the upload queue is clogged, _and_ we restart, we'll throw away the work that was stuck in the queue.
+
+What's probably most important though is the impact of this RFC on regular pageserver restart, e.g., during weekly deploys.
+In general, restart faces the problem of tenants that "take too long" to shut down.
+They are a problem because other tenants that shut down quickly are unavailble while we wait for the slow tenants to shut down.
+We currently allot 10 seconds for graceful shutdown until we SIGKILL the pageserver process (as per `pageserver.service` unit file).
+A longer budget would expose tenants that are done early to a longer downtime.
+A short budget would risk throwing away more work that'd have to be re-done after restart.
+In the context of this RFC, killing the process would mean losing the work that hasn't made it to S3.
+We can mitigate this problem as follows:
+0. initially, by accepting that we need to do the work again
+1. short-term, by introducing a read-only shutdown state for tenants that are fast to shut down; that state would be equivalent to the state of a tenant in hot standby / readonly mode.
+2. mid term, by not restarting pageserver in place, but using [*seamless tenant migration*](https://github.com/neondatabase/neon/pull/5029) to drain a pageserver's tenants before we restart it.
+
+## Side-Effects Of This Design
+
+* local `metadata` is basically reduced to a cache of which timelines exist for this tenant; i.e., we can avoid a `ListObjects` requests for a tenant's timelines during tenant load.
+
+## Limitations
+
+Multi-object changes that span multiple timelines aren't covered.
+We currently have:
+1. GC (but, it doesn't require atomicity across timelines, so, it doesn't really count)
+2. Timeline deletion (covered by `deleted_at` marker)
```

Comment:
This should probably expand on why this is about *timeline* deletion, because one might think it should read *tenant* deletion, because that's what first comes to mind when thtinking about work spanning multiple timelines. I think in the meeting we mentioned dependencies between different timelines making this multi-timeline?

Reply:
Hm, I think I'll just drop timeline & tenant deletion here. Both of them are really not "atomic layer map changes" but outright deletions of the layer map. No need to solve these. Will explain that.

Reply:
f51079302e8e7cace8ee6f2353763c147f024c07